h2o-core/src/main/java/hex/ModelBuilder.java

```diff
@@ -465,7 +465,7 @@ public Vec cv_AssignFold(int N) {
       }.doAll(2*N,Vec.T_NUM,folds_and_weights).outputFrame().vecs();
 
     if (_parms._keep_cross_validation_fold_assignment)
-      DKV.put(new Frame(Key.<Frame>make("cv_fold_assignment_" + _result.toString()), new String[]{"fold_assignment"}, new Vec[]{foldAssignment.makeCopy()}));
```

Comment:
Why do we don't need to make a copy? Can you please explain this change?

Reply:
my 2 cents here: foldAssignment is not manipulated after this line, so making a copy looks over-defensive coding.
Although, especially in relation with the following line, it could be make more explicit:
```
if (_parms._keep_cross_validation_fold_assignment)
      DKV.put(new Frame(Key.<Frame>make("cv_fold_assignment_" + _result.toString()), new String[]{"fold_assignment"}, new Vec[]{foldAssignment}));
else if( _parms._fold_column == null) foldAssignment.remove();
```
even wondering if last line should not be simply
`else foldAssignment.remove();`

Reply:
I would put it this way, I don't understand why .makeCopy() was there before. This code was added in scope of PUBDEV-2774 task. And there is no test that checks this makeCopy addition. We also don't have test in Vec for makeCopy method. I have following considerations:
1) Do we really need to copy? if there is a case when we remove (train) data and still need fold assignments? as I understand that fold assignment frame contains only one vector (with number of fold as values) and it is of no use without data.
2) I'm not sure that makeCopy works as expected.... i tried to add `Vec leaked = foldAssignment.makeCopy();` line and the test did not fail with leaking keys error.
3) I can't check that all the tests are fine without makeCopy() ... CI tests keep hang for different reasons. But If the tests are fine then I kind of can change the code however I want to :slightly_smiling_face:
Summary: making copy is expensive and we should try to avoid doing it whenever we can. Using fold assignments without data doesn't make much sense ( at least based on my current understanding).

Reply:
@seb-h2o @deil87 good points, I also don't see an obvious reason. 

One think to keep in mind is Scope, the Vec might be tracked by Scope and can be nuked after the Scope exits. One possibility is that the author wanted to prevent this - that would be my guess.

I don't junits will fail on this because they usually don't do anything after the model is built. But we might see something in runits/pyunits.

I will look into it as well.

Reply:
@michalkurka yes, you are right. foldassignment is being tracked by Scope from computeCrossValidation method. If author had an intention like you mentioned above then they should have used Scope.untrack on the copy.

Reply:
@deil87 yes, please add a test - that is the best way to figure this out

Reply:
@michalkurka done, please review the tests

Reply:
thanks, will review!